### PR TITLE
Cr 586

### PR DESCRIPTION
**Pull request by Aleksandr-codefresh**

### CODE REVIEW

File: __tests__/validator.unit.spec.js [hunk 1]
```diff
@@ -3603,8 +3603,7 @@ describe('Validate Codefresh YAML', () => {
                         'commands': ['jim', 'bob'],
                         'environment': ['key=value', 'key1=value¡'],
                         'fail_fast': true,
-                        'when': { branch: { only: ['master'] } },
-                        'create_file': 'yes'
```
Comment (Aleksandr-codefresh) [inline __tests__/validator.unit.spec.js:3607]: We do not have such property on a freestyle step.

File: schema/1.0/validator.js [hunk 2]
```diff
@@ -474,6 +475,17 @@ class Validator {
                 console.log(`Warning: no schema found for step type '${type}'. Skipping validation`);
                 continue; // eslint-disable-line no-continue
             }
+
+            if (type === Freestyle.getType()) {
```
Comment (Aleksandr-codefresh) [inline schema/1.0/validator.js:479]: Lets apply it for freestyle for now. And later expose it on other steps.

File: schema/1.0/validator.js [hunk 3]
```diff
@@ -484,6 +496,76 @@ class Validator {
     }
 
 
+    static _getStepSchemeProperties(stepSchema) {
+        const { children } = stepSchema.describe();
+        const renames = _.get(stepSchema, '_inner.renames', []).map(({ from }) => from);
```
Comment (Aleksandr-codefresh) [inline schema/1.0/validator.js:501]: Not sure if this is a correct way to get all renamed properties in a Schema.

File: yarn.lock [hunk 4]
```diff
@@ -2571,6 +2571,11 @@ joi@^9.2.0:
     moment "2.x.x"
     topo "2.x.x"
 
+js-levenshtein@^1.1.6:
```
Comment (Aleksandr-codefresh) [inline yarn.lock:2574]: This is an idea https://en.wikipedia.org/wiki/Levenshtein_distance

File: schema/1.0/steps/freestyle.js [hunk 5]
```diff
@@ -35,6 +35,7 @@ class Freestyle extends BaseSchema {
 
     getSchema() {
         const freestyleProperties = {
+            type: Joi.string().valid(Freestyle.getType()),
```
Comment (yaroslav-codefresh) [inline schema/1.0/steps/freestyle.js:38]: does it allows empty value?

Comment (Aleksandr-codefresh) [reply schema/1.0/steps/freestyle.js:38]: No it isn't. I will remove it

File: schema/1.0/validations/suggest-argument.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+const _ = require('lodash');
+const levenshtein = require('js-levenshtein');
+
+
+class SuggestArgumentValidation {
+
+    static get lengthThreshold() {
+        return 3;
```
Comment (yaroslav-codefresh) [inline schema/1.0/validations/suggest-argument.js:10]: Maybe should add some configuration from env vars?

File: schema/1.0/validations/suggest-argument.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+
+    static get distanceThreshold() {
+        return 5;
```
Comment (yaroslav-codefresh) [inline schema/1.0/validations/suggest-argument.js:15]: The same

File: schema/1.0/validations/suggest-argument.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+
+    static suggest(schema, argument, path) {
+        if (!schema) return '';
```
Comment (yaroslav-codefresh) [inline schema/1.0/validations/suggest-argument.js:20]: Maybe better to be null or undefined?

**Review by yaroslav-codefresh — APPROVED** [review]

File: schema/1.0/validations/suggest-argument.js (hunk 8 above)
Comment (Aleksandr-codefresh) [reply schema/1.0/validations/suggest-argument.js]: Fixed to null

**Pull request merged**